### PR TITLE
feat(integrations): database dependent testing

### LINKED ISSUE #27: Add support for database dependent testing
We want to be able to test all functionality, including those depending on the database. For this we need we need a strategy. To avoid the need for clearing and seeding the database for each test, we want to go for an in-memory solution. This requires a new implementation of the existing database interface.

### PR DESCRIPTION
**Pull request by JohnatMasking**
Fixes #27 

Changes proposed in this pull request:
- Adding in memory database implementation
- Setting MemoryDb as default implementation

@MaskingTechnology/comify

### CODE REVIEW

File: test/integrations/database/MemoryDb.spec.ts [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+            expect(result[1]).toEqual(expect.objectContaining(calzone));
+            expect(result[2]).toEqual(expect.objectContaining(hawai));
+            expect(result[3]).toEqual(expect.objectContaining(pepperoni));
+            expect(result[4]).toEqual(expect.objectContaining(margherita));
+        });
+
+        it('should sort the records by multiple fields same direction', async () =>
+        {
+            const query: RecordQuery = { };
+            const sort: RecordSort = { 'size': SortDirections.DESCENDING, 'name': SortDirections.DESCENDING };
```
Comment (BasMasking) [inline test/integrations/database/MemoryDb.spec.ts:354]: The result of this sort is non transparent. I expect this to be in visible in the fixture.

**Review by BasMasking — CHANGES_REQUESTED** [review]
One overall comment to a file that is not here. The mongodb.ts file is all written in lower case.

File: src/integrations/database/implementations/MemoryDb.ts
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:1]: name is adjusted.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+const LOGICAL_OPERATORS =
+{
+    [LogicalOperators.AND]: '&&',
+    [LogicalOperators.OR]: '||' 
+};
+
+export default class MemoryDb implements Database
+{
+    #memory?: Map<string, RecordData[]>;
+    recId: number = 0;
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:28]: Variable name changed.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    {
+        this.#memory = undefined;
+    }
+
+    async createRecord(type: string, data: RecordData): Promise<string>
+    {
+        const collection = this.#getCollection(type);
+        const id = this.#createId();
+        const record = {'id' : id, ...data};
+        collection.push(record);
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:46]: Empty line added.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+        return id; 
+    }
+
+    async readRecord(type: string, id: string, fields?: string[] | undefined): Promise<RecordData>
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:51]: Function prototype changed to option 1.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        }
+
+        for (const field of fields)
+        {
+            result[field] = data[field];
+        }
+
+        return result;
+    }
+
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:274]: Empty line removed.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        return collection;
+    }
+
+    #buildRecordData(data: RecordData, fields?: RecordField[]) : RecordData
+    { 
+        let result: RecordData = {};
+
+        if (fields === undefined)
+        {
+            result = { ...data };
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:262]: Statement combined and declaration made const.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+        if (index === -1)
+        {
+            throw new RecordNotFound();
+        }
+
+        collection.splice(index, 1);
+    }
+
+    async findRecord(type: string, query: QueryStatement, fields?: string[] | undefined, sort?: RecordSort | undefined): Promise<RecordData | undefined>
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:93]: Function declaration adjusted.

File: test/integrations/database/_fixtures/MemoryDb.fixture.ts
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/_fixtures/MemoryDb.fixture.ts:1]: All Dutch translated to English.

File: test/integrations/database/MemoryDb.spec.ts [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+import { describe, expect, it } from 'vitest';
+
+import { 
+    database,
+    RECORD_TYPE_PIZZA, RECORD_TYPE_FRUIT,
+    margherita, calzone, pepperoni, vegetario, hawai, margheritaId, peerId, appleId,
+    SortDirections, RecordNotFound, RecordNotUpdated, RecordData, RecordQuery, RecordSort
+} from './_fixtures/MemoryDb.fixture';
+
+describe('implementations/memorydb', () =>
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:11]: Name adjusted.

File: test/integrations/database/MemoryDb.spec.ts [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+            const result = await database.readRecord(RECORD_TYPE_PIZZA, margheritaId, [ 'id', 'folded' ]);
+
+            expect(Object.keys(result).length).toBe(2);
+            expect(result?.id).toBe(margheritaId);
+            expect(result?.folded).toBe(false);
+        });
+
+        it('should throw an error when a record is not found', async () =>
+        {
+            const result = database.readRecord(RECORD_TYPE_PIZZA, '12345678');
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:33]: INVALID_ID declared in the fixture.

File: test/integrations/database/MemoryDb.spec.ts [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+    {
+        it('should delete the record', async() =>
+        {
+            await database.deleteRecord(RECORD_TYPE_FRUIT, appleId);
+            const result = database.readRecord(RECORD_TYPE_FRUIT, appleId);
+
+            expect(result).rejects.toStrictEqual(new RecordNotFound());   
+        });
+
+        it('should throw an error when the record is not found', async () =>
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:49]: Text adjusted.

File: test/integrations/database/MemoryDb.spec.ts [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+        it('should update the record', async () =>
+        {
+            const updateData: RecordData = { land: 'Frankrijk' };
+            await database.updateRecord(RECORD_TYPE_FRUIT, peerId , updateData);
+            const result = await database.readRecord(RECORD_TYPE_FRUIT, peerId);
+
+            expect(result?.land).toBe(updateData.land);
+        });
+
+        it('should throw an error when record not found', async () =>
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:68]: Text adjusted.

File: test/integrations/database/MemoryDb.spec.ts [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+        {
+            const result = database.updateRecord(RECORD_TYPE_FRUIT, '12345678', {});
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:70]: INVALID_ID used.

File: test/integrations/database/_fixtures/MemoryDb.fixture.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+const RECORD_TYPE_FRUIT = 'fruit';
+
+const database = new MemoryDb();
+
+await database.connect('', 'test');
+
+const margherita: RecordData = { name: 'Margherita', size: 15, price: 12.00, folded: false };
+const margheritaId = await database.createRecord(RECORD_TYPE_PIZZA,  margherita);
+
+const apple: RecordData = { name: 'Appel', land: 'Belgie', bespoten: false };
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/_fixtures/MemoryDb.fixture.ts:17]: Different types are now grouped.

File: test/integrations/database/MemoryDb.spec.ts [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+
+            expect(result).rejects.toStrictEqual(new RecordNotFound());
+        });
+    });
+
+    describe('.updateRecord', () =>
+    {
+        it('should update the record', async () =>
+        {
+            const updateData: RecordData = { land: 'Frankrijk' };
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:61]: Land translated to country.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+        return id; 
+    }
+
+    async readRecord(type: string, id: string, fields?: string[] | undefined): Promise<RecordData>
+    {
+        const collection = this.#getCollection(type);
+        const record = collection.find(object => object.id === id);
+
+        if (record === undefined)
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:56]: undefined removed from interface.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    async findRecord(type: string, query: QueryStatement, fields?: string[] | undefined, sort?: RecordSort | undefined): Promise<RecordData | undefined>
+    {
+        const result = await this.searchRecords(type, query, fields, sort, 1, 0);
+
+        return result[0];
+    }
+
+    async searchRecords(type: string, query: QueryStatement, fields?: string[] | undefined, sort?: RecordSort | undefined, limit?: number | undefined, offset?: number | undefined): Promise<RecordData[]>
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:100]: Function declaration adjusted.

File: test/integrations/database/MemoryDb.spec.ts [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+            await database.updateRecord(RECORD_TYPE_FRUIT, peerId , updateData);
+            const result = await database.readRecord(RECORD_TYPE_FRUIT, peerId);
+
+            expect(result?.land).toBe(updateData.land);
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:65]: Updatedata declared in fixture.

File: test/integrations/database/MemoryDb.spec.ts [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+    });
+    
+    describe('.findRecord', () =>
+    {
+        it('should give only the first record found', async () =>
+        {
+            const query: RecordQuery = { };
+            const result = await database.findRecord(RECORD_TYPE_PIZZA, query);
+
+            expect(result).toEqual(expect.objectContaining(margherita));
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:83]: First item defined in fixture.

File: test/integrations/database/MemoryDb.spec.ts [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+        });
+
+        it('should find no record', async () =>
+        {
+            const query: RecordQuery = { name: { EQUALS: 'Romanos'} };
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:88]: Invalid pizza name defined in fixture

File: test/integrations/database/MemoryDb.spec.ts [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+
+            const result = await database.searchRecords(RECORD_TYPE_PIZZA, query);
+ 
+            expect(result[0]).toBe(undefined);
+        });
+    });
+
+    describe('.sortRecords', () =>
+    {
+        it('should sort records asending', async () =>
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:323]: typo corrected.

File: test/integrations/database/MemoryDb.spec.ts
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:1]: Please specify.

File: test/integrations/database/MemoryDb.spec.ts (hunk 1 above)
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:353]: All mutiple results specified in fixture.

**Review by BasMasking — CHANGES_REQUESTED** [review]

File: src/integrations/database/implementations/MemoryDb.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:94]: Findrecord also fixed.

File: test/integrations/database/MemoryDb.spec.ts
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:1]: Test with all options has been added.

File: src/integrations/database/implementations/MemoryDb.ts [hunk 21]
```diff
@@ -25,7 +25,7 @@ const LOGICAL_OPERATORS =
 export default class MemoryDb implements Database
 {
     #memory?: Map<string, RecordData[]>;
-    recId: number = 0;
+    recordId: number = 0;
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply src/integrations/database/implementations/MemoryDb.ts:28]: Typing removed.

File: test/integrations/database/_fixtures/MemoryDb.fixture.ts [hunk 22]
```diff
@@ -32,9 +35,41 @@ await database.createRecord(RECORD_TYPE_PIZZA, vegetario);
...
+const PIZZAS = {
+    MARGHERITA: margherita,
+    CALZONE: calzone,
+    PEPPERONI: pepperoni,
+    VEGETARIO: vegetario,
+    HAWAI: hawai
+};
+
+const IDS = {
+    MARGHERITAID: margheritaId,
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/_fixtures/MemoryDb.fixture.ts:47]: ID's changed.

File: test/integrations/database/MemoryDb.spec.ts [hunk 23]
```diff
@@ -384,18 +363,17 @@ describe('implementations/memorydb', () =>
                 const result = await database.searchRecords(RECORD_TYPE_PIZZA, query, undefined, undefined, 2);
                 
                 expect(result.length).toBe(2);
-                expect(result[0]).toEqual(expect.objectContaining(margherita));
-                expect(result[1]).toEqual(expect.objectContaining(calzone));
+                expect(result).toMatchObject(QUERY_RESULTS.LIMITED_BY_NUMBER);
+                expect(result[1]).toEqual(expect.objectContaining(PIZZAS.CALZONE));
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:367]: Old line removed.

File: test/integrations/database/MemoryDb.spec.ts [hunk 24]
```diff
@@ -3,34 +3,35 @@ import { describe, expect, it } from 'vitest';
 
 import { 
     database,
-    RECORD_TYPE_PIZZA, RECORD_TYPE_FRUIT,
-    margherita, calzone, pepperoni, vegetario, hawai, margheritaId, peerId, appleId,
+    RECORD_TYPE_PIZZA, RECORD_TYPE_FRUIT, INVALID_PIZZA_NAME, INVALID_ID, UPDATE_DATA,
+    PIZZAS, IDS, FIRST_PIZZA, QUERY_RESULTS,
     SortDirections, RecordNotFound, RecordNotUpdated, RecordData, RecordQuery, RecordSort
 } from './_fixtures/MemoryDb.fixture';
+import { ID } from '../../../src/integrations/database/definitions/constants';
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:10]: ID import moved to fixture.

File: test/integrations/database/MemoryDb.spec.ts [hunk 25]
```diff
@@ -58,16 +59,16 @@ describe('implementations/memorydb', () =>
     {
         it('should update the record', async () =>
         {
-            const updateData: RecordData = { land: 'Frankrijk' };
-            await database.updateRecord(RECORD_TYPE_FRUIT, peerId , updateData);
-            const result = await database.readRecord(RECORD_TYPE_FRUIT, peerId);
+            const updateData: RecordData = { country: UPDATE_DATA };
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/integrations/database/MemoryDb.spec.ts:62]: changed to UPDATE_COUNTRY

**Review by BasMasking — DISMISSED** [review]

**Review by BasMasking — APPROVED** [review]

**Pull request merged**